tests/experimental/datasources/test_integration_datasource.py

```diff
@@ -336,3 +351,148 @@ def _configure_and_run_data_assistant(
     checkpoint_result = checkpoint.run()
 
     return data_assistant_result, checkpoint_result
+
+
+@pytest.mark.integration
+@pytest.mark.parametrize(
+    ["n_rows", "fetch_all", "success"],
+    [
+        (None, False, True),
+        (3, False, True),
+        (7, False, True),
+        (-100, False, True),
+        ("invalid_value", False, False),
+        (1.5, False, False),
+        (True, False, False),
+        (0, False, True),
+        (200000, False, True),
+        (1, False, True),
+        (-50000, False, True),
+        (-5, True, True),
+        (0, True, True),
+        (3, True, True),
+        (50000, True, True),
+        (-20000, True, True),
+        (None, True, True),
+        (15, "invalid_value", False),
+    ],
+)
+def test_batch_head(
+    datasource_test_data: tuple[
+        AbstractDataContext, Datasource, DataAsset, BatchRequest
+    ],
+    fetch_all: bool | str,
+    n_rows: int | float | str | None,
+    success: bool,
+) -> None:
+    _, datasource, _, batch_request = datasource_test_data
+    batch_list: list[Batch] = datasource.get_batch_list_from_batch_request(
+        batch_request=batch_request
+    )
+    assert len(batch_list) > 0
+
+    # arbitrarily select the first batch for testing
+    batch: Batch = batch_list[0]
+    if success:
+        assert n_rows is None or isinstance(n_rows, int)
+        assert isinstance(fetch_all, bool)
+
+        execution_engine: ExecutionEngine = batch.data.execution_engine
+        execution_engine.batch_manager.load_batch_list(batch_list=[batch])
+
+        metrics_calculator = MetricsCalculator(execution_engine=execution_engine)
+
+        # define metric for the total number of rows
+        row_count_metric = MetricConfiguration(
+            metric_name="table.row_count",
+            metric_domain_kwargs={"batch_id": batch.id},
+            metric_value_kwargs=None,
+        )
```

Comment:
I think this is personal preference but if the test has a number of conditionals and branching paths, it may make sense to lose some value from parameterization in favor of easier readability/ease of debugging?

I think it may make sense to break up this test and the grouped asserts based on what we're actually testing but will defer to the core ZEP folks here

Reply:
Right now there are 36 test cases (will be 54 when Spark datasource is added). Almost all of the complexity you see here is because Spark returns something different than Pandas or SQL Alchemy. That would be one way to simplify, but then you lose the ability to ensure the 3 `ExecutionEngine`s are being tested similarly. We could also separate the column name logic from the row count logic if that makes readability better. My opinion is that the comments should give you enough information to debug.